<a href="https://colab.research.google.com/github/Kshitij-Panday/Aiml_Lab/blob/main/Kshitij_AIML_Lab2_EDA_Preprocessing_Workbook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Machine Learning Lab Experiment 2
# Topic: Exploratory Data Analysis, Data Preprocessing, and Data Visualization
# Prerequisite: Familiarity with Python, NumPy, Pandas

In [ ]:
# ================================
# Introduction
# ================================

In [ ]:
"""
This lab covers:
1. Exploratory Data Analysis (EDA)
2. Data Preprocessing techniques
3. Data Visualization
Each section includes explanations, code snippets, and tasks to reinforce learning.
"""

'\nThis lab covers:\n1. Exploratory Data Analysis (EDA)\n2. Data Preprocessing techniques\n3. Data Visualization\nEach section includes explanations, code snippets, and tasks to reinforce learning.\n'

In [ ]:
# ================================
# Importing Required Libraries
# ================================

In [ ]:
# Import necessary Python libraries for data manipulation, visualization, and preprocessing


import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer



In [ ]:
#================================
# Load the Dataset
# ================================



view the first few or last few rows of a DataFrame using the .head() or .tail() methods, respectively. You can specify the number of rows through the n argument (the default value is 5).

In [ ]:
"""
Use any CSV file like Titanic, Iris, or a generic CSV with categorical and numerical features.
"""
# Load dataset using pandas and display the first few rows



() Parentheses uses 1) Call a function 2) ordering expressions, 3)In conditions (recommended for clarity with and/or)

[] Square brackets uses 1) Access DataFrame/Series columns (Pandas)2)indexing/slicing 3)Create a list (mutable sequence)

{} Curly braces
1) Create a dictionary (key → value)
d = {"name": "Asha", "score": 82}

2) Create a set (unique values)
s = {1, 2, 2, 3}   

Empty {} is a dict, not a set:

{}          # empty dict
set()       # empty set

Quick memory rule

() → call / group / tuple

[] → list / indexing / slicing

{} → dictionary / set

view the first few or last few rows of a DataFrame using the .head() or .tail() methods, respectively. You can specify the number of rows through the n argument (the default value is 5).

In [ ]:
#Display first 5 rows


In [ ]:
# Display last 5 rows

12

In [ ]:
# Display column names and data types


In [ ]:
print(df.shape)

df.shape # (rows, columns)
df.columns# column names
df.dtypes # data type of each column
df.head() # first 5 rows
df.tail()# last 5 rows
df.sample(5)# random 5 rows


In [ ]:
df.describe()

In [ ]:
df.describe(include=[int])

In [ ]:
df.describe(exclude=[int])

In [ ]:
df.info()
df.shape

df.info()                         # non-null counts + dtypes

df.describe ()                     #  gives numeric summaries (mean, std, min/max).

df.describe(include="object")     # categorical/text summary

df.nunique()                      # unique count per column

df.describe(exclude=[int]) #exclude certain data types
df.describe(include=[int])
isolate specific data types in your summary output by using the include argument

In [ ]:
# Generate descriptive statistics


In [ ]:
df.describe(include=[int])
#df.describe(exclude=[int])

In [ ]:
#print("Shape (rows, columns):", df.shape)
#print("\nColumns:", list(df.columns))
#print("\nData types:\n", df.dtypes)

In [ ]:
# calculate mean of age column


np.float64(29.7)

In [ ]:
#slicing
df.columns[2:12]

df.isna().sum() # missing count per column

df.isna().mean()*100 # missing % per column

df[df.isna().any(axis=1)] # rows that have any missing value

In [1]:
# VIMP Check for missing values
#df.isna() checks which values are missing in the DataFrame and returns a table of True/False:
# returns True → value is missing (NA/NaN)
#False → value is present
#df.isna().sum() # missing count per column
df.isnull().head(7) #shows only the first 7 rows of that True/False table.
#df.isnull().sum()  # isnull() and isna() are same

NameError: name 'df' is not defined

In [ ]:
#Check how many ages are missing (good practice)


In [ ]:
#Pandas will automatically ignore NaN by default
#df["Age"].mean(skipna=True)



np.int64(177)

In [ ]:
df["Age"].head(10)

In [ ]:
#df[["Age", "Fare", "Survived"]].head()
#df[condition].head() → head of filtered data
#df[df["Age"] > 50].head(5)
#df.sort_values("Fare", ascending=False).head(5)   # top 5 highest fares
#df.sort_values("Age", ascending=True).head(5)     # 5 youngest (ignoring NaN positions)
#df.groupby("Gender")["Survived"].mean().head()
df.describe(include="all")

1)	Default describe() (numeric columns only)
df.describe() #Gives count, mean, std, min, 25%, 50%, 75%, max for numeric columns.
2)  with rounding (clean output)
df.describe().round(2)

3) Describe only one column (Series)
df["Age"].describe()
df["Fare"].describe()

4) Describe specific columns
df[["Age", "Fare", "SibSp"]].describe()

5) Include categorical columns too (include="all")
df.describe(include="all")
For text/categorical columns it shows:	count, unique, top, freq

6) Describe only categorical (object) columns
df.describe(include=["object"])
Useful for columns like Name, Ticket, Cabin, Embarked, etc.

7) Describe both numeric + category explicitly
df.describe(include=["number", "object"])

8) Exclude some types (example: exclude object)
df.describe(exclude=["object"])

9) Control percentiles shown (good for outliers)
df.describe(percentiles=[0.01, 0.05, 0.5, 0.95, 0.99])
Helps see extreme values and skewness (important for ML preprocessing).

10) Transpose describe for readability (columns become rows)
df.describe().T
Very helpful when there are many numeric columns.

Quick Titanic-focused examples
df[["Age", "Fare"]].describe(percentiles=[0.1, 0.5, 0.9]).round(2)
df.describe(include=["object"]).T



In [ ]:
#slice indexing:3 = start index (included)11 = end index (excluded)




In [ ]:
missing_count = df.isna().sum()
print(missing_count)

PassengerId      0
Survived         0
Pclass           0
Name             0
Gender           0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


In [ ]:
# Check for duplicate rows
#Duplicates can bias training.
df.duplicated().sum()
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [ ]:
#Target Variable Understanding: Survived (Text cell)
#In ML, we first study the target distribution (class balance).
df["Survived"].value_counts()
#0 usually means did not survive
#1 means survived

,count
Survived,
0,549
1,342


In [ ]:
#Survival percentage
#df["Survived"].value_counts(normalize=True)
#df["Survived"].value_counts(normalize=True) * 100
#Include missing values (if any)
df["Survived"].value_counts(dropna=False)
df["Survived"].value_counts().plot(kind="bar")

Exercise1 Task 1 (Text cell)

Task: Print only the Age column and show:

first 8 values

how many missing values are in Age

In [ ]:
#solution Task1


Task 2
Task: Find the top 5 most common tickets (Ticket values).

In [ ]:
# solution Task2


Task: Find the survival rate by Gender (as percentage).

Step-by-step explanation
1) Groups the data into separate groups based on Gender:
one group for Female and one group for Male

use groupby() and use Gender attribute

 df.groupby("Gender")

 2) From each group, select only the Survived column.["Survived"]



3)Compute the average of Survived for each gender.
 .mean()

Survival rate by Gender (in %)
Survived is usually: 1 =survived
0 = not survived
So the average (mean) of 0/1 becomes a rate.

Example: if among females [1, 1, 0, 1]
Mean = (1+1+0+1)/4 = 0.75 → 75% survival rate



In [ ]:
#Task: Find the survival rate by Gender (as percentage).
grouped = df.groupby("Gender")
#print(grouped.groups.keys()) #See group names (Male/Female)
#print(grouped.get_group("male").head()) #Get one group and print it (e.g., Male)
#print (grouped.shape ())
#(df.groupby("Gender")["Survived"].mean() * 100).round(2)
for gender, data in grouped:
    print("\nGender:", gender)
    print("Rows:", len(data))
    print(data.head(3))

value_counts() counts how many times each unique value appears in a column (Series).

Most common use in EDA to check category distribution and class balance.

In [ ]:
# View value counts of categorical columns (e.g., Gender, Embarked)
df['Gender'].value_counts()
df['Embarked'].value_counts()
#df["Gender"].value_counts(normalize=True) * 100
#for col in ["Gender", "Embarked", "Pclass"]:
#print("\n", col)
#print(df[col].value_counts(dropna=False))


In [ ]:
# ================================
# 2. Handling Missing Data
# ================================

In [ ]:
df.isna().sum()

In [ ]:
df1=df.copy()
df1.head (30)

1) Copy an existing DataFrame (most common)
df2 = df.copy()

Use this when you want a new DataFrame without changing the original.

2) Create from a dictionary (columns → lists)
df3 = pd.DataFrame({
    "name": ["Asha", "Ravi", "Neha"],
    "score": [82, 75, 90],
    "passed": [1, 1, 1]
})

3) Create a DataFrame from selected columns of another
df4 = df[["Age", "Fare", "Survived"]].copy()

4) Create a DataFrame by filtering rows
df5 = df[df["Age"] > 60].copy()

5) Create a DataFrame by dropping columns
df6 = df.drop(columns=["Cabin", "Ticket"]).copy()

6) Create an empty DataFrame (then fill later)
df7 = pd.DataFrame(columns=["A", "B", "C"])

In [ ]:
df6=df.copy()
df6.shape
df6.columns
df10=df.drop(columns=["PassengerId","Name","Cabin"])
df10.shape


In [ ]:
df11=df.drop(columns=["Cabin", "Ticket"]).copy()
df11.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Gender', 'Age', 'SibSp',
       'Parch', 'Fare', 'Embarked'],
      dtype='object')

In [ ]:
# Apply imputation strategy to fill missing numerical values (e.g., Age)
num_imputer = SimpleImputer(strategy = 'median')
#num_imputer.fit(df[['Age']])
#df['Age'] = num_imputer.transform(df[['Age']])


df['Age'] = num_imputer.fit_transform(df[['Age']])

In [ ]:
df['Age'].isnull().sum()

np.int64(0)

In [ ]:
df['Embarked']

In [ ]:
missing = df.isna().sum()
missing_pct = (df.isna().mean()*100).round(2)
print(pd.DataFrame({"missing": missing, "missing_%": missing_pct}).sort_values("missing", ascending=False))

Column-by-column strategy (ML perspective)
1) PassengerId (0 missing) Role: Identifier only  Action: Drop before ML (not useful for prediction)

2) Survived (0 missing)Role: Target/Label
Action: Keep as y

3) Pclass (0 missing)
Role: Important feature (ordinal category 1/2/3)
Action: Keep; later encode as numeric/one-hot (your choice)

4) Name (0 missing)
Role: Raw text (not directly ML-friendly)
Best use: Extract Title (Mr/Mrs/Miss/etc.) then drop Name
Action: Keep temporarily for feature engineering

5) Gender (0 missing)
Role: Strong categorical predictor
Action: Keep; encode later (male/female → 0/1 or one-hot)

6) Age (177 missing ~19.87%)
Role: Important numeric feature
Best strategy: Impute + add missing flag

Create Age_missing = 1 if Age is missing else 0
Fill missing Age using a smart median:
Better: median by (Pclass, Gender) or (Title, Pclass)
Simpler: overall median

Impute Age by median of (Pclass, Gender) + add Age_missing

7) SibSp (0 missing)
Role: Useful numeric feature
Action: Keep; engineer FamilySize = SibSp + Parch + 1, IsAlone

8) Parch (0 missing)
Role: Useful numeric feature
Action: Keep; used in FamilySize

9) Ticket (0 missing)
Role: High-cardinality text (many unique values)
Action: Usually drop for baseline ML

10) Fare (0 missing)
Role: Important numeric feature (often skewed/outliers)
Action: Keep


11) Cabin (687 missing ~77.10%)
Role: Very sparse text feature
Best strategy: Don’t use raw Cabin directly
Choose one approach:
Option A (simplest baseline):Drop Cabin entirely

Option B (ML-friendly + valuable):Create CabinKnown (0/1)

Create Deck = first letter of cabin (A/B/C/...) and fill missing as "Unknown"
Then drop raw Cabin

Recommended for teaching ML thinking:
Use Deck + CabinKnown, drop Cabin

12) Embarked (2 missing ~0.22%)
Role: Small categorical feature
Action: Fill missing with mode (most frequent value), then encode

In [ ]:
df1=df.copy()
df1.head (30)

In [ ]:
cat_imputer = SimpleImputer(strategy='most_frequent')
df['Embarked'] = df['Embarked'].astype('category')
df['Embarked'] = cat_imputer.fit_transform(df[['Embarked']]).ravel()

In [ ]:
df['Embarked'].isnull().sum()

np.int64(0)

In [ ]:
# Apply imputation strategy to fill missing categorical values (e.g., Embarked)

In [ ]:
# ================================
# 3. Removing Duplicate Entries
# ================================

In [ ]:
# Remove duplicate rows from the dataset





In [ ]:
# ================================
# 4. Outlier Detection and Treatment
# ================================

In [ ]:
# Use the IQR method to detect outliers in numerical features (e.g., Fare)
Q1 =df['Fare'].quantile(0.25)
Q2 =df['Fare'].quantile(0.50)
Q3 =df['Fare'].quantile(0.75)
IQR =Q3-Q1



In [ ]:
# Optionally remove or transform detected outliers





In [ ]:
# ================================
# 5. Encoding Categorical Variables
# ================================

In [ ]:
# Apply label encoding to Gender column





In [ ]:
# Apply one-hot encoding to categorical variables (e.g., Embarked)





In [ ]:
# ================================
# 6. Feature Scaling
# ================================

In [ ]:
# Apply standardization or normalization to numerical features (e.g., Age, Fare)





In [ ]:
# ================================
# 7. Feature Engineering
# ================================

In [ ]:
# Create new features (e.g., FamilySize = SibSp + Parch + 1)





In [ ]:
# ================================
# 8. Feature Selection
# ================================

In [ ]:
# Use filter methods like SelectKBest to select top features




In [ ]:
# Use wrapper methods like RFE to rank features





In [ ]:
# ================================
# 9. Splitting Dataset
# ================================

In [ ]:
# Split the dataset into training and test sets using train_test_split





In [ ]:
# ================================
# 10. Visualization
# ================================

In [ ]:
# Plot histograms to understand distributions (e.g., Age)





In [ ]:
# Create scatter plots to explore relationships (e.g., Fare vs Age colored by Survived)






In [ ]:
# Display a heatmap to show feature correlations







In [ ]:
# ================================
# Lab Exercises
# ================================
"""
1. Load a different dataset (e.g., Iris or any from UCI repo).
2. Perform similar EDA steps on it.
3. Try different imputation strategies.
4. Detect and visualize outliers in at least two features.
5. Create 2 new features based on domain knowledge.
6. Apply both standardization and normalization and compare.
7. Use feature selection and plot feature importance.
8. Divide the dataset and plot the training vs testing distributions.
9. Visualize data using pair plots or violin plots.
10. Write a short report on insights found from the EDA.
"""

'\n1. Load a different dataset (e.g., Iris or any from UCI repo).\n2. Perform similar EDA steps on it.\n3. Try different imputation strategies.\n4. Detect and visualize outliers in at least two features.\n5. Create 2 new features based on domain knowledge.\n6. Apply both standardization and normalization and compare.\n7. Use feature selection and plot feature importance.\n8. Divide the dataset and plot the training vs testing distributions.\n9. Visualize data using pair plots or violin plots.\n10. Write a short report on insights found from the EDA.\n'

In [ ]:
# END OF LAB 1 NOTEBOOK


In [ ]:
from google.colab import files
import io

uploaded = files.upload()


Saving titanic.csv to titanic.csv


In [ ]:
import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer

df = pd.read_csv('titanic.csv')

print(df.to_string())

     PassengerId  Survived  Pclass                                                                                Name  Gender    Age  SibSp  Parch              Ticket      Fare            Cabin Embarked
0              1         0       3                                                             Braund, Mr. Owen Harris    male  22.00      1      0           A/5 21171    7.2500              NaN        S
1              2         1       1                                 Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.00      1      0            PC 17599   71.2833              C85        C
2              3         1       3                                                              Heikkinen, Miss. Laina  female  26.00      0      0    STON/O2. 3101282    7.9250              NaN        S
3              4         1       1                                        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.00      1      0              113803   53.1000       

In [ ]:
print(df.shape)

(891, 12)


In [ ]:
df.describe(include=[int])

,PassengerId,Survived,Pclass,SibSp,Parch
count,891.000000,891.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,0.523008,0.381594
std,257.353842,0.486592,0.836071,1.102743,0.806057
min,1.000000,0.000000,1.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,0.000000,0.000000
50%,446.000000,0.000000,3.000000,0.000000,0.000000
75%,668.500000,1.000000,3.000000,1.000000,0.000000
max,891.000000,1.000000,3.000000,8.000000,6.000000


In [ ]:
df.describe(include=[float])

,Age,Fare
count,714.000000,891.000000
mean,29.699118,32.204208
std,14.526497,49.693429
min,0.420000,0.000000
25%,20.125000,7.910400
50%,28.000000,14.454200
75%,38.000000,31.000000
max,80.000000,512.329200


In [ ]:
df.describe(exclude=[int])

,Name,Gender,Age,Ticket,Fare,Cabin,Embarked
count,891,891,714.000000,891,891.000000,204,889
unique,891,2,NaN,681,NaN,147,3
top,"Dooley, Mr. Patrick",male,NaN,347082,NaN,G6,S
freq,1,577,NaN,7,NaN,4,644
mean,NaN,NaN,29.699118,NaN,32.204208,NaN,NaN
std,NaN,NaN,14.526497,NaN,49.693429,NaN,NaN
min,NaN,NaN,0.420000,NaN,0.000000,NaN,NaN
25%,NaN,NaN,20.125000,NaN,7.910400,NaN,NaN
50%,NaN,NaN,28.000000,NaN,14.454200,NaN,NaN
75%,NaN,NaN,38.000000,NaN,31.000000,NaN,NaN


In [ ]:
df.describe(exclude=[float])

,PassengerId,Survived,Pclass,Name,Gender,SibSp,Parch,Ticket,Cabin,Embarked
count,891.000000,891.000000,891.000000,891,891,891.000000,891.000000,891,204,889
unique,NaN,NaN,NaN,891,2,NaN,NaN,681,147,3
top,NaN,NaN,NaN,"Dooley, Mr. Patrick",male,NaN,NaN,347082,G6,S
freq,NaN,NaN,NaN,1,577,NaN,NaN,7,4,644
mean,446.000000,0.383838,2.308642,NaN,NaN,0.523008,0.381594,NaN,NaN,NaN
std,257.353842,0.486592,0.836071,NaN,NaN,1.102743,0.806057,NaN,NaN,NaN
min,1.000000,0.000000,1.000000,NaN,NaN,0.000000,0.000000,NaN,NaN,NaN
25%,223.500000,0.000000,2.000000,NaN,NaN,0.000000,0.000000,NaN,NaN,NaN
50%,446.000000,0.000000,3.000000,NaN,NaN,0.000000,0.000000,NaN,NaN,NaN
75%,668.500000,1.000000,3.000000,NaN,NaN,1.000000,0.000000,NaN,NaN,NaN


In [ ]:
df.duplicated().sum()
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0
